File: BackTracking/Subsets.cs
// Problem: https://leetcode.com/problems/subsets/
// Author: Dung Nguyen Tien (Chris)

public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        var res = new List<IList<int>>();
        
        backtrack(nums, res, new List<int>(), 0);
        
        return res;
    }
    
    void backtrack(int[] nums, List<IList<int>>res, List<int> tmp, int start) {
        res.Add(new List<int>(tmp));
        
        for (var i = start; i < nums.Length; i++) {
            tmp.Add(nums[i]);
            backtrack(nums, res, tmp, i + 1);
            tmp.RemoveAt(tmp.Count - 1);
        }
    }
}

File: Javascript/Subsets.js
/**
 * Problem: https://leetcode.com/problems/subsets/submissions/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    var res = [];
    
    backTrack(res, [], nums, 0);
    
    return res;
};

function backTrack(res, tmp, nums, start) {
    res.push(tmp.slice(0));
    
    for (var i = start; i < nums.length; i++) {
        // pick the first solution
        tmp.push(nums[i]);
        // backtrack
        backTrack(res, tmp, nums, i + 1);
        tmp.pop();
    }
}